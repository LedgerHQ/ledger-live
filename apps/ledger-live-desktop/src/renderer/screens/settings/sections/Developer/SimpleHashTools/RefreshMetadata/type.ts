export enum RefreshOption {
  nft = "nft",
  contract = "contract",
}
