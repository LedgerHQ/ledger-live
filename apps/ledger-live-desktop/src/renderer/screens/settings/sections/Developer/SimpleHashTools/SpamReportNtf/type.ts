export enum ReportOption {
  individual = "individual",
  contract = "contract",
  collection = "collection",
}
