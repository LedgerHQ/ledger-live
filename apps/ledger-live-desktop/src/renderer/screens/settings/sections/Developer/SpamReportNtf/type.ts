export enum ReportOption {
  individual = "individual",
  contract = "contract",
  collection = "collection",
}

export type SelectOption = { value: string; label: string };
