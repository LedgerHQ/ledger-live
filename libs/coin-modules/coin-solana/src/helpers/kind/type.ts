import { OperationType } from "@ledgerhq/types-live";

export type OperationTypeMapper = {
  kind: string;
  operationType: OperationType;
};
