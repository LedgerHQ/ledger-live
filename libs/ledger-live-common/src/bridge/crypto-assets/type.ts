import { TokenCurrency } from "@ledgerhq/types-cryptoassets";

export type ERC20Token = {
  parentCurrencyId: string;
  token: string;
  ticker: string;
  magnitude: number;
  name: string;
  ledgerSignature: string;
  contractAddress: string;
  disableCountervalue: false;
  delisted: boolean;
};

export type SPLToken = {
  id: string;
  network: string;
  name: string;
  symbol: string;
  address: string;
  decimals: number;
};

export type CryptoAssetsStore = {
  findTokenByAddress(address: string): TokenCurrency | undefined;
  getTokenById(id: string): TokenCurrency;
  addTokens(list: (TokenCurrency | undefined)[]): void;
  convertERC20(token: ERC20Token): TokenCurrency | undefined;
  findTokenById(id: string): TokenCurrency | undefined;
  findTokenByAddressInCurrency(address: string, currencyId: string): TokenCurrency | undefined;
  findTokenByTicker(ticker: string): TokenCurrency | undefined;
  convertSplTokens(token: SPLToken): TokenCurrency;
};
