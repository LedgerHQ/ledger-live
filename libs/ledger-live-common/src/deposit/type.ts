import { CryptoOrTokenCurrency } from "@ledgerhq/types-cryptoassets";

export type MappedAsset = {
  $type: "Token" | "Coin";
  ledgerId: string;
  providerId: string;
  name: string;
  ticker: string;
  network?: string;
  contract?: string;
  status: string;
  reason: null;
  data: {
    img: string;
    marketCapRank: number | null;
  };
  ledgerCurrency?: CryptoOrTokenCurrency;
};

export type GroupedCurrency = {
  providerId: string;
  currenciesByNetwork: MappedAsset[];
};

export type CurrenciesByProviderId = {
  currenciesByNetwork: CryptoOrTokenCurrency[];
  providerId: string;
};
