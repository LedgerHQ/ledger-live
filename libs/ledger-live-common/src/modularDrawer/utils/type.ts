import { CounterValuesState } from "@ledgerhq/live-countervalues/lib/types";
import { CryptoOrTokenCurrency, Currency } from "@ledgerhq/types-cryptoassets";
import { AccountLike } from "@ledgerhq/types-live";
import { ReactNode } from "react";
import { Observable } from "rxjs";
import { WalletAPIAccount } from "../../wallet-api/types";
import { CurrenciesByProviderId } from "../../deposit/type";
import { EnhancedModularDrawerConfiguration } from "../../wallet-api/ModularDrawer/types";

export type ApyType = "NRR" | "APY" | "APR";

export type AssetType = {
  name: string;
  ticker: string;
  id: string;
  leftElement?: React.ReactNode;
  rightElement?: React.ReactNode;
};

export type ProviderBalanceAsset = {
  id: string;
  name: string;
  ticker: string;
  balance: string;
  fiatValue: string;
  sortValue: number;
};

export type ProviderBalanceResultsMap = Map<string, ProviderBalanceAsset>;

export type AssetWithBalance = AssetType & {
  sortValue?: number;
};

export type UseBalanceDeps = () => {
  flattenedAccounts: AccountLike[];
  discreet: boolean;
  state: CounterValuesState;
  counterValueCurrency: Currency;
  locale: string;
};

export type BalanceUI = { balance?: string; fiatValue?: string };
export type CreateBalanceItem = (x: BalanceUI) => React.ReactNode;

export type CreateAccountsCountAndApy = (args: {
  label?: string;
  value?: number;
  type?: ApyType;
}) => ReactNode;

export type NetworkWithCount = CryptoOrTokenCurrency & {
  leftElement?: ReactNode;
  count: number;
};

export type AccountDataItem = {
  asset: CryptoOrTokenCurrency;
  label: string;
  count: number;
};

export type AccountModuleParams = {
  assets: CryptoOrTokenCurrency[];
  accounts$?: Observable<WalletAPIAccount[]>;
};

export type CreateAccountsCount = (args: { label: string }) => ReactNode;

export type UseAccountData = (params: AccountModuleParams) => AccountDataItem[];

export type NetworkHookParams = {
  assets: CryptoOrTokenCurrency[];
  networks: CryptoOrTokenCurrency[];
  selectedAssetId: string;
  currenciesByProvider: CurrenciesByProviderId[];
  accounts$?: Observable<WalletAPIAccount[]>;
};
export type NetworkHook = (params: NetworkHookParams) => Array<CryptoOrTokenCurrency & Network>;

export type NetworkConfigurationDeps = {
  useAccountData: UseAccountData;
  accountsCount: CreateAccountsCount;
  accountsCountAndApy: CreateAccountsCountAndApy;
  useBalanceDeps: UseBalanceDeps;
  balanceItem: CreateBalanceItem;
};

export type LeftElementKind = "numberOfAccounts" | "numberOfAccountsAndApy" | "undefined";
export type RightElementKind = "balance" | "undefined";

export type Network = {
  name: string;
  id: string;
  ticker: string;
  leftElement?: React.ReactNode;
  rightElement?: React.ReactNode;
};

export type CreateNetworkConfigurationHookProps = {
  networksConfig: EnhancedModularDrawerConfiguration["networks"];
  currenciesByProvider?: CurrenciesByProviderId[];
  selectedAssetId: string;
  accounts$?: Observable<WalletAPIAccount[]>;
};

type Props = {
  assetsConfiguration: EnhancedModularDrawerConfiguration["assets"];
  currenciesByProvider?: CurrenciesByProviderId[];
};

export type AssetConfigurationDeps = {
  ApyIndicator: (args: { value: number; type: ApyType }) => ReactNode;
  useBalanceDeps: UseBalanceDeps;
  balanceItem: (asset: { fiatValue?: string; balance?: string }) => ReactNode;
};

export type CreateAssetConfigurationHook = (
  AssetConfigurationDeps: AssetConfigurationDeps,
) => (props: Props) => (assets: CryptoOrTokenCurrency[]) => (CryptoOrTokenCurrency & AssetType)[];
