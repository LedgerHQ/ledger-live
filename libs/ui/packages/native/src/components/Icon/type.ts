import React from "react";

export type IconProps = { size?: number; color?: string };
export type IconType = React.ComponentType<IconProps>;

export type NewIconProps = { size?: "XS" | "S" | "M" | "L" | "XL"; color?: string; style?: object };
export type NewIconType = React.ComponentType<NewIconProps>;

export type IconOrElementType = IconType | React.ReactElement;
export type NewIconOrElementType = NewIconType | React.ReactElement;
