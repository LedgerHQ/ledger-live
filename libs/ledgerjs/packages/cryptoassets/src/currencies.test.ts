import type { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import {
  listFiatCurrencies,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
} from "./fiats";
import {
  listTokens,
  findTokenById,
  findTokenByAddressInCurrency,
} from "./tokens";
import {
  listCryptoCurrencies,
  hasCryptoCurrencyId,
  getCryptoCurrencyById,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByTicker,
  findCryptoCurrencyByKeyword,
  registerCryptoCurrency,
  cryptocurrenciesById,
} from "./currencies";

test("can get currency by coin type", () => {
  expect(getCryptoCurrencyById("bitcoin")).toMatchObject({
    id: "bitcoin",
    name: "Bitcoin",
  });
  expect(getCryptoCurrencyById("litecoin")).toMatchObject({
    id: "litecoin",
    name: "Litecoin",
  });
  expect(hasCryptoCurrencyId("bitcoin")).toBe(true);
  expect(hasCryptoCurrencyId("")).toBe(false);
  expect(() => getCryptoCurrencyById("")).toThrow();
  expect(hasCryptoCurrencyId("_")).toBe(false);
  expect(() => getCryptoCurrencyById("_")).toThrow();
});

// Unique list in case of dup in the currency list
const currencies = Array.from(new Set(Object.values(cryptocurrenciesById)));
const duplicatedTickers: Set<string> = new Set();
currencies.reduce((acc, curr) => {
  if (acc.includes(curr.ticker)) {
    duplicatedTickers.add(curr.ticker);
  }
  acc.push(curr.ticker);
  return acc;
}, [] as string[]);

currencies.forEach((c) => {
  test(`should find ${c.name} by id`, () => {
    expect(findCryptoCurrencyById(c.id)).toEqual(c);
  });

  test(`should find ${c.name} by ticker`, () => {
    try {
      expect(findCryptoCurrencyByTicker(c.ticker)).toEqual(c);
    } catch (e) {
      // Should throw only if the ticker is not duplicated or if it's not a testnet (see conditions in `cryptocurrenciesByTicker`)
      if (!duplicatedTickers.has(c.ticker) && !c.isTestnetFor) {
        throw e;
      }
    }
  });

  test(`should find ${c.name} by scheme`, () => {
    expect(findCryptoCurrencyByScheme(c.scheme)).toEqual(c);
  });

  c.keywords?.forEach((k) => {
    test(`should find ${c.name} with keyword ${k}`, () => {
      expect(findCryptoCurrencyByKeyword(k)).toEqual(c);
    });
  });
});

test("[LEGACY TEST] can find currency", () => {
  const bitcoinMatch = {
    id: "bitcoin",
    name: "Bitcoin",
  };
  const ethereumMatch = {
    id: "ethereum",
    name: "Ethereum",
  };

  expect(findCryptoCurrency((c) => c.name === "Bitcoin")).toMatchObject(
    bitcoinMatch
  );
  expect(findCryptoCurrencyById("bitcoin")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByKeyword("btc")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByTicker("BTC")).toMatchObject(bitcoinMatch);
  expect(findCryptoCurrencyByScheme("bitcoin")).toMatchObject(bitcoinMatch);

  expect(findCryptoCurrencyById("ethereum")).toMatchObject(ethereumMatch);
  expect(findCryptoCurrencyByKeyword("eth")).toMatchObject(ethereumMatch);
  expect(findCryptoCurrencyByTicker("ETH")).toMatchObject(ethereumMatch);
  expect(findCryptoCurrencyByScheme("ethereum")).toMatchObject(ethereumMatch);

  expect(findCryptoCurrencyById("_")).toBe(undefined);
  expect(findCryptoCurrencyByKeyword("_")).toBe(undefined);
  expect(findCryptoCurrencyByTicker("_")).toBe(undefined);
  expect(findCryptoCurrencyByScheme("_")).toBe(undefined);
});

test("there are some dev cryptocurrencies", () => {
  const all = listCryptoCurrencies(true);
  const prod = listCryptoCurrencies();
  expect(all).not.toBe(prod);
  expect(all.filter((a) => !a.isTestnetFor)).toMatchObject(prod);
  expect(all.length).toBeGreaterThan(prod.length);
});

test("there are some terminated cryptocurrencies", () => {
  const all = listCryptoCurrencies(false, true);
  const supported = listCryptoCurrencies();
  expect(all).not.toBe(supported);
  expect(all.filter((a) => !a.terminated)).toMatchObject(supported);
  expect(all.length).toBeGreaterThan(supported.length);
});

test("all cryptocurrencies match (by reference) the one you get by id", () => {
  for (const c of listCryptoCurrencies()) {
    expect(c).toBe(getCryptoCurrencyById(c.id));
  }
});

test("there is no testnet or terminated coin by default", () => {
  expect(listCryptoCurrencies(false, false)).toBe(listCryptoCurrencies());
  expect(listCryptoCurrencies(true, true).length).toBeGreaterThan(
    listCryptoCurrencies().length
  );

  for (const c of listCryptoCurrencies()) {
    expect(!c.terminated).toBe(true);
    expect(!c.isTestnetFor).toBe(true);
  }
});

test("all cryptocurrencies have at least one unit", () => {
  for (const c of listCryptoCurrencies()) {
    expect(c.units.length).toBeGreaterThan(0);
  }
});

test("fiats list is always the same", () => {
  expect(listFiatCurrencies()).toEqual(listFiatCurrencies());
});

test("fiats list elements are correct", () => {
  const tickers = {};

  for (const fiat of listFiatCurrencies()) {
    expect(fiat.ticker).toBeTruthy();
    expect(typeof fiat.ticker).toBe("string");
    expect(tickers[fiat.ticker]).toBeFalsy();
    expect(fiat.units.length).toBeGreaterThan(0);
    const unit = fiat.units[0];
    expect(unit.code).toBeTruthy();
    expect(typeof unit.code).toBe("string");
    expect(unit.name).toBeTruthy();
    expect(typeof unit.name).toBe("string");
    expect(unit.magnitude).toBeGreaterThan(-1);
    expect(typeof unit.magnitude).toBe("number");
    tickers[fiat.ticker] = unit;
  }
});

test("tokens are correct", () => {
  expect(listTokens().length).toBeGreaterThan(0);

  for (const token of listTokens()) {
    expect(token.ticker).toBeTruthy();
    expect(typeof token.id).toBe("string");
    expect(typeof token.name).toBe("string");

    if (token.ledgerSignature) {
      expect(typeof token.ledgerSignature).toBe("string");
    }

    expect(typeof token.tokenType).toBe("string");
    expect(typeof token.parentCurrency).toBe("object");
    expect(hasCryptoCurrencyId(token.parentCurrency.id)).toBe(true);
    expect(typeof token.ticker).toBe("string");
    expect(token.units.length).toBeGreaterThan(0);
    const unit = token.units[0];
    expect(unit.code).toBeTruthy();
    expect(typeof unit.code).toBe("string");
    expect(unit.name).toBeTruthy();
    expect(typeof unit.name).toBe("string");
    expect(unit.magnitude).toBeGreaterThan(-1);
    expect(typeof unit.magnitude).toBe("number");

    if (token.compoundFor) {
      const t = findTokenById(token.compoundFor);
      expect(typeof t).toBe("object");
    }
  }
});

test("findTokenByAddressInCurrency", () => {
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "bsc"
    )
  ).toMatchObject({
    id: "bsc/bep20/1inch_token",
  });
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "ethereum"
    )
  ).toMatchObject({
    id: "ethereum/erc20/1inch_token",
  });
  expect(findTokenByAddressInCurrency("0x0", "bsc")).toBe(undefined);
  expect(
    findTokenByAddressInCurrency(
      "0x111111111117dC0aa78b770fA6A738034120C302",
      "tron"
    )
  ).toBe(undefined);
});

test("fiats list is sorted by ticker", () => {
  expect(
    listFiatCurrencies()
      .map((fiat) => fiat.ticker)
      .join(",")
  ).toEqual(
    listFiatCurrencies()
      .map((fiat) => fiat.ticker)
      .sort((a, b) => (a > b ? 1 : -1))
      .join(",")
  );
});

test("testnet currencies must also set disableCountervalue to true", () => {
  expect(
    listCryptoCurrencies(true)
      .filter((c) => c.isTestnetFor)
      .filter((c) => !c.disableCountervalue)
      .map((c) => c.id)
  ).toEqual([]);
});

test("can get fiat by coin type", () => {
  expect(getFiatCurrencyByTicker("USD").units[0]).toMatchObject({
    magnitude: 2,
  });
  expect(getFiatCurrencyByTicker("EUR").units[0]).toMatchObject({
    magnitude: 2,
  });
  // this is not a fiat \o/
  expect(() => getFiatCurrencyByTicker("USDT").units[0]).toThrow();
  expect(hasFiatCurrencyTicker("USD")).toBe(true);
  expect(hasFiatCurrencyTicker("USDT")).toBe(false);
});

test("all USDT are countervalue enabled", () => {
  const tokens = listTokens().filter(
    (t) => t.ticker === "USDT" && !t.parentCurrency.isTestnetFor
  );
  expect(tokens.map((t) => t.id).sort()).toMatchSnapshot();
  expect(tokens.every((t) => t.disableCountervalue === false)).toBe(true);
});

test("Ethereum family convention: all ethereum testnet coins must derivate on the same cointype as the testnet it's for (e.g. ethereum ropsten is on 60)", () => {
  expect(
    listCryptoCurrencies()
      .filter(
        (e) =>
          e.family === "ethereum" && // ethereum family
          e.isTestnetFor && // is a testnet coin
          e.coinType !== getCryptoCurrencyById(e.isTestnetFor).coinType // it must use same coinType as the mainnet coin
      )
      .map((e) => e.id) // to get a nice error if it fails
  ).toEqual([]);
});

test("can register a new coin externally", () => {
  const coinId = "mycoin";
  expect(() => getCryptoCurrencyById("mycoin")).toThrow(
    `currency with id "${coinId}" not found`
  );
  const mycoin = {
    type: "CryptoCurrency",
    id: coinId,
    coinType: 8008,
    name: "MyCoin",
    managerAppName: "MyCoin",
    ticker: "MYC",
    countervalueTicker: "MYC",
    scheme: "mycoin",
    color: "#ff0000",
    family: "mycoin",
    units: [
      {
        name: "MYC",
        code: "MYC",
        magnitude: 8,
      },
      {
        name: "SmallestUnit",
        code: "SMALLESTUNIT",
        magnitude: 0,
      },
    ],
    explorerViews: [
      {
        address: "https://mycoinexplorer.com/account/$address",
        tx: "https://mycoinexplorer.com/transaction/$hash",
        token: "https://mycoinexplorer.com/token/$contractAddress/?a=$address",
      },
    ],
  };
  registerCryptoCurrency(coinId, mycoin as CryptoCurrency);
  expect(getCryptoCurrencyById(coinId)).toEqual(mycoin);
});
