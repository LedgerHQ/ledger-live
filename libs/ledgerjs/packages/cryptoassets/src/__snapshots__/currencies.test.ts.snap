// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all USDT are countervalue enabled 1`] = `
[
  "algorand/asa/312769",
  "bsc/bep20/binance-peg_bsc-usd",
  "elrond/esdt/555344542d663863303863",
  "ethereum/erc20/usd_tether__erc20_",
  "polygon/erc20/(pos)_tether_usd",
  "tron/trc20/tr7nhqjekqxgtci8q8zy4pl8otszgjlj6t",
]
`;
